<!DOCTYPE html>

<html xmlns:th="http://www.thymeleaf.org">

<body>
    <div th:fragment="proto-tree">
        <form class="col-md-8 my-textarea full-pipe">
            <div class="row pipelines-examples">
                <div>
                    <a class="pipelines-example1">Example1</a>
                </div>
                <div>
                    <a class="pipelines-example2">Example2</a>
                </div>
                <div>
                    <a class="pipelines-example3">Example3</a>
                </div>
            </div>
            <div class="row">
                <textarea id="first-file-area" class="form-control" rows="12"> </textarea>
            </div>
            <br />
            <div class="row">
                <button id="addSecondArea" class="btn btn-primary btn-md proto-tree-button" type="button">
                    Add Second Area
                </button>
                <label class="btn btn-primary btn-md proto-tree-button">
                    Choose file <input id="first-file" type="file" style="display: none" multiple/>
                </label>
                <span class='label label-info' id="first-file-info"></span>
                <button id="GoAsync" class="btn btn-primary btn-md pull-right proto-tree-button" type="button">
                    <span>SUBMIT</span>
                </button>
            </div>
            <div class="row">
                <span id="first-area-message">Wrong format. The fasta format is expected to begin with '>' sign.</span>
                <span id="both-areas-message">To start the analysis you need to choose all the necessary files or paste the sequences in the areas.</span>
                <span id="one-area-message">To start the analysis you need to choose a file with protein sequences or paste the sequences in the area above.</span>
                <span id="malformed-fasta">There should be at least 3 protein sequences to start the analysis.</span>
                <span id="cannot-have-ids-and-seqs">You can't both provide sequences and check 'Retrieve sequences'</span>
            </div>
            <div class="row active-second-area trend-text-colour">
                <p>Full-length protein sequences go here.</p>
            </div>
            <div class="row fetch-fromIds1">
                <div class="checkbox">
                    <label><input id="fetch-fromIds-value1" type="checkbox">Retrieve sequences (NCBI and MiST)</label>
                </div>
            </div>
            <br /> <br />
        </form>
        <form class="col-md-8 my-textarea second-area">
            <div class="row">
                <textarea id="second-file-area" class="form-control" rows="12"> </textarea>
            </div>
            <br />
            <div class="row">
                <label class="btn btn-primary btn-md proto-tree-button">
                    Choose file <input id="second-file" type="file" style="display: none" multiple/>
                </label>
                <span class='label label-info' id="second-file-info"></span>
            </div>
            <div class="row">
                <span id="second-area-message">Wrong format. The fasta format is expected to begin with '>' sign.</span>
                <span id="malformed-fasta-second">There should be at least 3 protein sequences to start the analysis.</span>
            </div>
            <div class="row active-second-area trend-text-colour">
                <p>Fragments of protein sequences go here.</p>
            </div>
            <br /> <br />
        </form>
        <div class="col-md-8 my-textarea partial-pipe">
            <p class="trend-text-colour">Protein names in the provided files should be matched.</p>
            <p class="trend-text-colour">The following combinations of the files can be provided depending on the task:</p>
            <ul class="trend-text-colour">
                <li>
                    File with the tree + File with the alignment + File with initial protein sequences.
                    <p>Protein features will be identified and mapped on the tree, the alignment will be reordered according to the tree leaves.</p>
                </li>
                <li>
                    File with the tree + File with the alignment.
                    <p>The alignment will be reordered according to the tree leaves.</p>
                </li>
                <li>
                    File with the tree + File with initial protein sequences.
                    <p>Protein features will be identified and mapped on the tree.</p>
                </li>
                <li>
                    Just a file with the tree. "Retrieve sequences (NCBI and MiST)" box should be checked. Protein identifiers should be placed at the beginning of proteins names on the tree (See Help).
                    <p>Corresponding proteins will be fetched from <a href="https://www.ncbi.nlm.nih.gov/" target="_blank">NCBI</a> and/or <a href="https://mistdb.com/" target="_blank">MiST</a> databases, protein features will be identified and mapped on the tree.</p>
                </li>
            </ul>
            <div class="row partial-pipe-file partial-opt">
                <div class="partial-pipe-interior">
                    <label class="btn btn-primary btn-md proto-tree-button">
                        Choose file with protein sequences (FASTA) <input id="sequence-file" type="file" style="display: none" multiple/>
                    </label>
                    <span class='label label-info' id="sequence-file-info"></span>
                </div>
            </div>
            <div class="row">
                <span id="first-area-message-partialP">Wrong format. The fasta format is expected to begin with '>' sign.</span>
                <span id="malformed-fasta-partialP">There should be at least 3 protein sequences to start the analysis.</span>
            </div>
            <div class="row partial-pipe-file partial-opt">
                <div class="partial-pipe-interior">
                    <label class="btn btn-primary btn-md proto-tree-button">
                        Choose file with alignment (FASTA) <input id="alignment-file" type="file" style="display: none" multiple/>
                    </label>
                    <span class='label label-info' id="alignment-file-info"></span>
                </div>
            </div>
            <div class="row">
                <span id="second-area-message-partialP">Wrong format. The fasta format is expected to begin with '>' sign.</span>
                <span id="malformed-fasta-second-partialP">There should be at least 3 protein sequences to start the analysis.</span>
            </div>
            <div class="row">
                <div class="partial-pipe-file">
                    <div class="partial-pipe-interior">
                        <label class="btn btn-primary btn-md proto-tree-button">
                            Choose file with tree (Newick) <input id="tree-file" type="file" style="display: none" multiple/>
                        </label>
                        <span class='label label-info' id="tree-file-info"></span>
                    </div>
                </div>
                <div>
                    <span id="partial-pipeline-message">To start the analysis both a file with phylogenetic tree and a file with protein sequences and/or a file with the alignment should be provided. <p>Another option is to provide a file with phylogenetic tree and check "Retrieve sequences" below.</p></span>
                    <span id="malformed-newick">Malformed newick tree. Please correct the tree.</span>
                </div>
                <button id="GoAsyncPartial" class="btn btn-primary btn-md pull-right proto-tree-button" type="button">
                    <span>SUBMIT</span>
                </button>
            </div>
            <div class="row fetch-fromTree-value">
                <div class="checkbox">
                    <label><input id="fetch-fromTree-value" type="checkbox">Retrieve sequences (NCBI and MiST)</label>
                </div>
            </div>
        </div>

    </div>

    <div th:fragment="result">
        <div class="progress">
            <div id="pipelineProgressBar" class="progress-bar progress-bar-striped active" aria-valuenow="0" role="progressbar" aria-valuemin="0" aria-valuemax="100" style="width:0%">
            </div>
        </div>
        <div class="col-md-11 wait-for-it">
            <div id="check-later">
                You may want to copy and save this URL to check the status of your job later:
                <a id="check-by-url" href=""></a> You may also bookmark this page.
            </div>
            <div id="result-stage"></div>
        </div>
        <div class="col-md-11 no-such-biojob">
            No such job found
        </div>
        <div class="col-md-11 server-error">
            Error happened while processing your input. Please check the input and try again.
            <br/>
            If you are sure your input is correct please contact us providing the details of your analysis and the input.
        </div>
        <div class="col-md-12 text-center result-container">
            <div class="result-download-buttons">
                <a id="results-load" class="btn btn-info btn-md proto-tree-button" href="#">
                    Tree (SVG) <span class="glyphicon glyphicon-download-alt "> </span>
                </a>
                <a id="tree-load" class="btn btn-info btn-md proto-tree-button" href="#">
                    Tree (Newick) <span class="glyphicon glyphicon-download-alt"> </span>
                </a>
                <a id="alignment-load" class="btn btn-info btn-md proto-tree-button" href="#">
                    Alignment (FASTA) <span class="glyphicon glyphicon-download-alt"> </span>
                </a>
                <a id="features-load" class="btn btn-info btn-md proto-tree-button" href="#">
                    Protein Features (JSON) <span class="glyphicon glyphicon-download-alt"> </span>
                </a>
                <a id="cdhit-clusters-load" class="btn btn-info btn-md proto-tree-button" href="#">
                    CD-HIT clusters (.clstr) <span class="glyphicon glyphicon-download-alt"> </span>
                </a>
            </div>
        </div>
        <div class="col-md-12 my-result">
            <div style="display: none">
                <span id="jobId" th:text="${jobId}"></span>
            </div>
            <div id="svgContainer"></div>
        </div>
    </div>

    <div th:fragment="navigator">
        <div class="navigator">
            <button id="full-pipeline" class="btn btn-primary btn-md pipeline-button-selected" type="button">
                Full Pipeline
            </button>
            <button id="partial-pipeline" class="btn btn-primary btn-md pipeline-button" type="button">
                Partial Pipeline
            </button>
        </div>
    </div>

    <div th:fragment="proto-tree-filter">
        <div class="col-md-2 panel panel-default">
            <div class="panel-body">
                <h2 class="text-primary info-header">Set options</h2>
                <div class="form-group full-pipe alignment-and-tree">
                    <form id="first">
                        <div class="cd-hit">
                            <h3 class="text-info" >Redundancy (CD-HIT):</h3>
                            <div class="checkbox">
                                <label><input id="reduce-redundancy" type="checkbox" >Reduce sequence redundancy</label>
                            </div>
                            <input type="number" min="0.4" step="0.01" max="1" value="0.9" id="cd-hit" name="cd-hit" class="form-control input-md"/>
                        </div>
                        <h3 class="text-info" >Alignment (Mafft):</h3>
                        <div class="checkbox">
                            <label><input id="do-align" type="checkbox" >Align</label>
                        </div>
                        <span class="aln-alg">Alignment Algorithm:</span>
                        <select id="alignment-alg" class="form-control input-md aln-alg">
                            <option value="--retree 2 --maxiterate 0">FFT-NS-2</option>
                            <option selected='selected' value="--retree 2 --maxiterate 1000">FFT-NS-i</option>
                            <option value="--maxiterate 1000 --localpair">L-INS-i</option>
                            <option value="--maxiterate 1000 --genafpair">E-INS-i</option>
                            <option value="--maxiterate 1000 --globalpair">G-INS-i</option>
                        </select>
                        <h3 class="text-info">Tree Building:</h3>
                        <div class="radio tree-soft">
                            <label>
                                <input id="fast-tree-soft" name="tree-soft" type="radio" value="FastTree" checked="checked" />FastTree
                            </label>
                        </div>
                        <div class="radio tree-soft">
                            <label>
                                <input id="mega-soft" name="tree-soft" type="radio" value="Mega"/>Mega
                            </label>
                        </div>
                        <div class="fast-tree">
                            <span>A.A. Substitution Model:</span>
                            <select id="subst-model-ft" class="form-control input-md">
                                <option selected='selected' value="jtt">JTT+CAT</option>
                                <option value="wag">WAG+CAT</option>
                                <option value="lg">LG+CAT</option>
                            </select>
                            <div class="ft-options">
                                <span>Phylogeny Test:</span>
                                <select id="phylo-test-ft" class="form-control input-md">
                                    <option value="nosupport">None</option>
                                    <option selected='selected' value="sh">Shimodaira-Hasegawa test</option>
                                    <option value="nome">Minimum-evolution bootstrap</option>
                                </select>
                            </div>
                            <div class="number-of-replicates-ft">
                                <span>Number of replicates:</span>
                                <input min="50" max="1000" value="100" id="number-of-replicates-ft" name="number-of-ft-replicates" type="number" class="form-control input-md"/>
                            </div>
                            <div class="checkbox">
                                <label><input id="pseudo-counts-ft" type="checkbox">Use pseudocounts (recommended for highly gapped sequences)</label>
                            </div>
                        </div>
                        <div class="mega-tree">
                            <span>Tree Building Method:</span>
                            <select id="tree-method" class="form-control input-md">
                                <option selected='selected' value="JN">Neighbour Joining</option>
                                <option value="ML">Maximum Likelihood</option>
                                <option value="ME">Minimum Evolution</option>
                            </select>
                            <div class="ml-options">
                                <span>A.A. Substitution Model:</span>
                                <select id="ml-subst-model" class="form-control input-md">
                                    <option selected='selected' value="jtt">Jones-Taylor-Thornton (JTT) model</option>
                                    <option value="pm">Poisson model</option>
                                    <option value="eim">Equal input model</option>
                                    <option value="dm">Dayhoff model</option>
                                    <option value="dmf">Dayhoff model with Freqs. (F+)</option>
                                    <option value="jttf">JTT with Freqs. (F+) model</option>
                                    <option value="wm">WAG model</option>
                                    <option value="wmf">WAG with Freqs. (F+) model</option>
                                    <option value="lg">LG model</option>
                                    <option value="lgf">LG with Freqs. (F+) model</option>
                                    <option value="grm">General Reversible Mitochondrial (mtREV)</option>
                                    <option value="grmf">mtREV with Freqs. (F+) model</option>
                                    <option value="grc">General Reversible Chloroplast (cpREV)</option>
                                    <option value="grcf">cpREV with Freqs. (F+) model</option>
                                    <option value="grt">General Reversible Transcriptase model (rtREV)</option>
                                    <option value="grtf">rtREV with Freqs. (F+) model</option>
                                </select>
                            </div>
                            <div class="nj-and-me-options">
                                <span>A.A. Substitution Model:</span>
                                <select id="nj_me-subst-model" class="form-control input-md">
                                    <option selected='selected' value="jtt">Jones-Taylor-Thornton (JTT) model</option>
                                    <option value="nd">No. of differences</option>
                                    <option value="pd">p-distance</option>
                                    <option value="pm">Poisson model</option>
                                    <option value="eim">Equal input model</option>
                                    <option value="dm">Dayhoff model</option>
                                </select>
                            </div>
                            <span>A.A. Substitution Rate:</span>
                            <select id="subst-rate" class="form-control input-md">
                                <option selected='selected' value="ur">Uniform Rates</option>
                                <option value="gd">Gamma Distributed (G)</option>
                                <option value="ir">Has Invariant Sites (I)</option>
                                <option value="gir">Gamma Distributed With Invariant Sites (G+I)</option>
                            </select>
                            <div class="ml-options">
                                <span>Initial tree for ML:</span>
                                <select id="initial-tree-ml" class="form-control input-md">
                                    <option selected='selected' value="njBio">Make initial tree automatically (Default - NJ/BioNJ)</option>
                                    <option value="mp">Make initial tree automatically (Maximum parsimony)</option>
                                    <option value="nj">Make initial tree automatically (Neighbor joining)</option>
                                    <option value="bioNj">Make initial tree automatically (BioNJ)</option>
                                </select>
                            </div>
                            <span>Gaps and Missing Data:</span>
                            <select id="gaps-missing" class="form-control input-md">
                                <option selected='selected' value="compDel">Complete deletion</option>
                                <option value="partDel">Partial deletion</option>
                                <option value="pairDel">Pairwise deletion</option>
                            </select>
                            <div class="cov-cutoff">
                                <span>Site Coverage Cutoff:</span>
                                <input value="95" id="site-cov-cutoff" name="site-cov-cutoff" type="text" class="form-control input-sm"/>
                            </div>
                            <div class="nj-and-me-options">
                                <span>Phylogeny Test:</span>
                                <select id="nj_me-phylo-test" class="form-control input-md">
                                    <option selected='selected' value="none">None</option>
                                    <option value="bm">Bootstrap method</option>
                                    <option value="ib">Interior-branch test</option>
                                </select>
                            </div>
                            <div class="ml-options">
                                <span>Phylogeny Test:</span>
                                <select id="ml-phylo-test" class="form-control input-md">
                                    <option selected='selected' value="none">None</option>
                                    <option value="bm">Bootstrap method</option>
                                </select>
                            </div>
                            <div class="number-of-replicates">
                                <span>Number of replicates:</span>
                                <input min="50" max="500" value="100" id="number-of-replicates" name="number-of-replicates" type="number" class="form-control input-md"/>
                            </div>
                        </div>
                    </form>
                </div>
                <span id="isFullPipeline">true</span>
                <div class="form-group">
                    <h3 class="text-info">Protein Features:</h3>
                    <div class="checkbox">
                        <label><input id="do-predict-features" type="checkbox" >Identify Features</label>
                    </div>
                    <form id="second">
                        <span class="predict-features">Algorithm:</span>
                        <div class="radio predict-features">
                            <label>
                                <input id="hmmscan" name="dom-prediction-program" type="radio" value="hmmscan" checked="checked" />HMMER3
                            </label>
                        </div>
                        <div class="radio predict-features">
                            <label>
                                <input id="rpsblast" name="dom-prediction-program" type="radio" value="rpsblast"  />RPSBLAST
                            </label>
                        </div>
                        <div id="hmmer-db" class="predict-features">
                            <span>HMMER Databse:</span>
                            <select class="form-control input-md hmmer-db">
                                <option id="pfam-only" selected='selected' value="pfam-only">Pfam 35</option>
                                <option id="pfam_and_mist"  value="pfam_and_mist">Pfam 35 + MiST + TREND</option>
                            </select>
                            <span>E-value threshold:</span>
                            <input id="evalue-threshold-hmmer" type="number" min="0" max="1" step="0.01" value="0.1" name="evalue-threshold" class="form-control input-md"/>
                            <span>Domain Score:</span>
                            <input id="probability" type="number" min="0" value="20" class="form-control input-md"  >
                        </div>
                        <div id="rpsblast-db" class="predict-features">
                            <span>RPSBLAST Database:</span>
                            <select class="form-control input-md rpsblast-db">
                                <option id="cdd" selected='selected' value="cdd">CDD SuperSet</option>
                                <option id="cdd_ncbi" value="cdd_ncbi">CDD NCBI</option>
                                <option id="pfam" value="pfam">Pfam</option>
                                <option id="cog" value="cog">COG</option>
                                <option id="kog" value="kog">KOG</option>
                                <option id="smart" value="smart">SMART</option>
                                <option id="prk" value="prk">PRK</option>
                                <option id="tigr" value="tigr">TIGR</option>
                            </select>
                            <span>E-value threshold:</span>
                            <input id="evalue-threshold-rpsblast" type="number" min="0" max="1" step="0.01" value="0.1" name="evalue-threshold" class="form-control input-md"/>
                        </div>
                        <br/>
                        <div class="switcher predict-features">
                            <span class="switch-text">Low Complexity:</span>
                            <label class="switch">
                                <input id="lc-value" type="checkbox"/>
                                <span class="slider round"></span>
                            </label>
                        </div>
                        <hr/>
                        <div class="switcher">
                            <span class="switch-text">Enumerate:</span>
                            <label class="switch">
                                <input id="enumerate-value" type="checkbox"/>
                                <span class="slider round"></span>
                            </label>
                        </div>
                    </form>
                    <button id="filter-clear" class="btn btn-primary btn-md pull-right proto-tree-button" type="button">
                        <span>RESET</span>
                    </button>
                </div>
            </div>
        </div>
    </div>
</body>