<!DOCTYPE html>

<html xmlns:th="http://www.thymeleaf.org">

<div th:fragment="contact">
    <div class="row">
        <div class="col-md-12">
            <h2>Support & feedback</h2>
            <div class="contact">
                Send your comments or bug reports to: comp.bioverse@gmail.com
            </div>

        </div>
    </div>
</div>

<div th:fragment="credits">
    <div class="row">
        <div class="col-md-12 credits-list">
            <div class="credits">
                <a href="https://mafft.cbrc.jp/alignment/software/" target="_blank">MAFFT</a>
            </div>
            <div class="credits">
                <a href="http://www.microbesonline.org/fasttree/" target="_blank">FastTree</a>
            </div>
            <div class="credits">
                <a href="https://www.megasoftware.net/" target="_blank">MEGA</a>
            </div>
            <div class="credits">
                <a href="http://hmmer.org/" target="_blank">HMMER</a>
            </div>
            <div class="credits">
                <a href="http://www.cbs.dtu.dk/services/TMHMM/" target="_blank">TMHMM</a>
            </div>
            <div class="credits">
                <a href="http://weizhongli-lab.org/cd-hit/" target="_blank">CD-HIT</a>
            </div>
            <div class="credits">
                <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi" target="_blank">BLAST</a>
            </div>
            <div class="credits">
                <a href="https://www.ncbi.nlm.nih.gov/Structure/cdd/cdd.shtml" target="_blank">CDD</a>
            </div>
            <div class="credits">
                <a href="https://pfam.xfam.org/" target="_blank">Pfam</a>
            </div>
            <div class="credits">
                <a href="https://mistdb.com/" target="_blank">MiST</a>
            </div>
            <div class="credits">
                <a href="http://etetoolkit.org/" target="_blank">ETE Toolkit</a>
            </div>
            <div class="credits">
                 <a href="https://github.com/guyleonard/jsPhyloSVG" target="_blank">jsPhyloSVG</a>
            </div>


        </div>
    </div>
</div>
<div th:fragment="about">
    <div class="row about-div">
        <div class="col-md-11 home-description">
            <h2 id="moto">Searching for regularities in protein evolution... </h2>
        </div>
    </div>
    <div class="row about-div">
        <div class="col-md-4 home-description">
            <h3 class="h3-about">Align protein sequences and build a tree</h3>
            Provide protein sequences and start the analysis or use our embedded id parser.
            Use a variety of MAFFT alignment algorithms and MEGA tree building options.
            Reorder the alignment according to the tree leaves.
        </div>
        <div class="col-md-4 home-description">
            <h3 class="h3-about">Identify protein domain architectures</h3>
            Identify protein domains, transmembrane and low-complexity regions.
            Make use of Pfam, CDD and MiST models.
        </div>
        <div class="col-md-4 home-description">
            <h3 class="h3-about">Identify gene neighborhoods</h3>
            Identify gene neighborhoods and operons for 125000+ genomes.
            Set your own distance threshold between the genes in operons.
            Set the number of neighboring genes to visualize.
        </div>
    </div>
    <div class="row about-div">
        <div class="col-md-4 home-description">
            <h3 class="h3-about">Cluster genes</h3>
            Cluster genes based on identified shared domains of encoded proteins.
            Customize the number of not shared domains for genes to be clustered.
        </div>
        <div class="col-md-4 home-description">
            <h3 class="h3-about">Visualize results on the tree</h3>
            Combine the tree and identified domains or gene neighborhoods in one interactive picture.
            Explore domains and gene neighborhoods.
            Take advantage of —Åross-links with Pfam, CDD, NCBI and MiST.
            Save SVG figure and detailed results of each step.
        </div>
        <div class="col-md-4 home-description">
            <h3 class="h3-about">Bring your own data</h3>
            If you have your own tree upload it and run subsequent steps.
            You can optionally upload your alignment to reorder it.
        </div>
    </div>
</div>

<div th:fragment="help">
    <div class="col-md-12 help-div">
        <h2 class="info-header">Analysis</h2>
        <div>
            <p>Study of evolutionary history and function(s) of a protein is an intriguing but not a trivial task.</p>
            <p>The computational analysis usually starts with collecting homologues proteins.</p>
            <p>Consequent steps include: </p>
            <ul>
                <li class="info">
                    creating multiple sequence alignment;
                </li>
                <li class="info">
                    building phylogenetic tree;
                </li>
                <li class="info">
                    identifying protein domains;
                </li>
                <li class="info">
                    identifying transmembrane regions and other protein features;
                </li>
                <li class="info">
                    determining gene neighborhoods.
                </li>
            </ul>
            <p>
                The next crucial step is analysis of sequence conservation, protein features and gene neighborhoods in the context of phylogenetic clustering.
            </p>
            <p>
                Such comprehensive analysis is time-consuming and error-prone.
            </p>
            <p>
                To help scientists in analysis of their data we have created a protein analysis hub that automates these steps
                and furthermore provides additional instruments.
            </p>
        </div>
    </div>
    <div class="col-md-12 help-div">
        <div id="domains-href" class="info-header">
            <h2 data-toggle="collapse" data-target="#domains-info">
                Domains <i class="fa fa-chevron-down chevron-info"></i>
            </h2>
        </div>
        <div id="domains-info" class="collapse">
            <div>
                <p class="text-head3"> Full pipeline mode. </p>
                <p> There are two ways the pipeline can be used in this mode:</p>
                <h4 class="text-head4" style="font-weight: bold">
                    1. Predicting protein features and building phylogenetic tree using full-length sequences.
                </h4>
                <div class="indented-text">
                    <p>
                        If you want to use this option just paste your sequences or upload a file and click <span class="proto-tree-text"> SUBMIT </span> button.
                    </p>
                    <p>
                        You can also provide a list of NCBI or MiST protein ids/locus tags instead of protein sequences, and our id analyzer will extract the sequences from  NCBI and MiST.
                    </p>
                    <p>
                        Just don't forget to tick <span class="proto-tree-text">Retrieve sequences (NCBI and MiST)</span>.
                    </p>
                    <p>
                        If you have your own alignment that you already prepared and edited you may skip the alignment step of the pipeline and use your alignment instead. Just paste your alignment and uncheck <span class="proto-tree-text">Align</span> option.
                    </p>
                    <p>
                        If you want to use TREND just to align your sequences and build a tree you may skip domains identification step unchecking <span class="proto-tree-text">Identify Features</span> option.
                    </p>
                </div>
                <h4  class="text-head4" style="font-weight: bold">
                    2. Building phylogenetic tree using fragments of protein sequences and identifying protein features using full-length sequences.
                </h4>
                <div class="indented-text">
                    <p>
                        If you want to use this option:
                    </p>
                    <p>
                        a) Click <span class="proto-tree-text">Add second area </span>button. New query area will appear.
                    </p>
                    <p>
                        b) Put full-length sequences into the first area (or choose a file). Protein features will be identified using these sequences.
                    </p>
                    <p>
                        c) Put fragments of sequences into the second area (or choose a file). Protein sequences in this area will be used for alignment and building phylogentic tree.
                    </p>
                    <p>
                        Important: sequence names in two areas should be identical.
                    </p>
                    <p>
                        If you downloaded sequences from the same database or were changing sequence names consistently in all the files this will happen naturally.
                    </p>
                </div>
            </div>
            <div>
                <p class="text-head3"> Partial pipeline mode. </p>
                <p>
                    If you have your own phylogenetic tree and protein sequences used to produce the tree use this mode.
                </p>
                <p>
                    <span style="font-weight: bold"> Important </span>: sequence names in all the files should be matched.
                </p>
                <p>
                    If you downloaded sequences from the same database or were changing sequence names consistently in all the files this will happen naturally.
                </p>
                <div><p style="font-weight:bold">First way:</p>
                    <div class="indented-text">
                        <p>
                            a) In <span class="proto-tree-text">Domains</span> section click <span class="proto-tree-text">Partial Pipeline</span> button.
                        </p>
                        <p>
                            b) Upload protein sequences used to build the tree (<span class="proto-tree-text">Choose file with protein sequences (fasta)</span> button).
                        </p>
                        <p>
                            c) Optional, if you want the alignment to be ordered according to the tree leaves order, upload your alignment (<span class="proto-tree-text">Choose file with alignment (fasta)</span> button).
                        </p>
                        <p>
                            d) Upload the tree in <a href="https://en.wikipedia.org/wiki/Newick_format" target="_blank">newick</a> format (<span class="proto-tree-text">Choose file with tree (Newick)</span> button).
                        </p>
                        <p>
                            If you just want to reorder sequences in your alignment according to the tree leaves just upload the alignment and tree.
                        </p>
                    </div>
                </div>
                <div><p style="font-weight:bold">Second way:</p>
                    <div class="indented-text">
                        <p>
                            a) Put at the beginning of the tree leaves names NCBI/MiST protein ids/locus tags separated from the rest of the names by
                            space, underscore ( _ ) or vertical bar (|).
                        </p>
                        <p>
                            b) Tick <span class="proto-tree-text">Retrieve sequences (NCBI and MiST)</span>. Our id analyzer will extract the sequences from NCBI and MiST.
                        </p>
                        <p>
                            c) And then just upload the tree in <a href="https://en.wikipedia.org/wiki/Newick_format" target="_blank">newick</a> format (<span class="proto-tree-text">Choose file with tree (Newick)</span> button) and start the analysis
                        </p>
                        <p>
                            If you download sequences from NCBI or MiST and use them to build the tree, the ids will be naturally at the beginning of sequences and you don't have to do anything.
                        </p>

                    </div>
                </div>
            </div>
            <div>
                <br/>
                <p>
                    As a result of running <span class="proto-tree-text">Domains</span> pipeline a phylogenetic tree combined with interactive protein features will be produced.
                </p>
                <p>
                    Details of the analysis can be explored by clicking on protein features. To zoom in/zoom out use mouse wheel.
                </p>
                <p>
                    All the produced data is downloadable.
                </p>
            </div>
        </div>
    </div>

    <div class="col-md-12 help-div">
        <hr/>
        <div id="neib-href" class="info-header">
            <h2 data-toggle="collapse" data-target="#neighborhoods-info">
                Neighborhoods <i class="fa fa-chevron-down chevron-info"></i>
            </h2>
        </div>
        <div id="neighborhoods-info" class="collapse">
            <div>
                <p>
                    You can cluster genes based on the shared domains of the encoded proteins and visualize the clusters and gene neighborhoods on phylogenetic tree running our neighborhoods analysis pipeline.
                </p>
                <p>
                    Protein names in the file with the sequences or the tree should start with protein identifiers
                    separated from the rest of the name by space, underscore ( _ ) or vertical bar (|).
                </p>
                    <p>
                        Suitable identifiers:
                    </p>
                    <p class="indented-text">
                        a) <a href="https://www.ncbi.nlm.nih.gov/refseq/about/nonredundantproteins/" target="_blank">NCBI RefSeq Id</a>;
                    </p>
                    <p class="indented-text">
                        b) <a href="https://www.ncbi.nlm.nih.gov/genomes/locustag/Proposal.pdf" target="_blank">Locus tag</a>, either old or new;
                    </p>
                    <p class="indented-text">
                        b) <a href="https://mistdb.com/" target="_blank">MiST Id</a>.
                    </p>
                <p>
                    The analyzer can be run in three ways:
                </p>
                <p>
                    1) by pasting protein sequences (or uploading corresponding file)
                </p>
                <p>
                    2) by pasting phylogenetic tree in newick format (or uploading corresponding file)
                </p>
                <p>
                    3) by pasting a list of NCBI/MiST protein ids/locus tags (or uploading corresponding file). Don't forget to check <span class="proto-tree-text">Retrieve sequences (NCBI and MiST)</span>
                </p>
                <p>
                    4) by pasting an alignment in <a href="https://en.wikipedia.org/wiki/FASTA_format" target="_blank">FASTA format</a> (or uploading corresponding file)
                </p>
                <p>
                    Bear in mind that the refSeq Ids are not organism specific. If you want to explore the neighborhoods of genes from a particular organism you should provide locus tags or MiST Ids.
                </p>
                <br/>
                <p>
                    <strong>'Operon tolerance'</strong> parameter is a distance in nucleotides between neighboring genes to consider them as being encoded in one operon.
                </p>
                <p>
                    <strong>'Not shared domains tolerance'</strong> parameter is a number of not shared domains between any two proteins that is allowed
                    for corresponding genes sill to be considered as members of the same cluster.
                </p>
                <p>
                    <strong>'Number of neighboring genes on one side (max 15)'</strong> parameter is a number of neighboring genes on each side of a gene of interest. Maximum is 15, i.e.
                    30 neighboring genes in sum will be displayed.
                </p>
            </div>
            <div>
                <br/>
                <p>
                    As a result of running <span class="proto-tree-text">Neighborhoods</span> pipeline a phylogenetic tree combined with interactive gene neighborhoods will be produced.
                </p>
                <p>
                    Genes that belong to the same cluster will be colored in the same color. Genes that belong to the same operon will have the borders of the same color.
                </p>
                <p>
                    Details of the neighboring genes can be explored by hovering mouse pointer. To zoom in/zoom out use mouse wheel.
                </p>
                <p>
                    All the produced data is downloadable.
                </p>
            </div>
        </div>
        <br/>
    </div>
    <div class="col-md-12 help-div">
        <br/>
        <div class="info-header">
            <h2>Browser compatibility</h2>
        </div>
        <table class="table">
            <thead>
            <tr>
                <th>OS</th>
                <th>Version</th>
                <th>Chrome</th>
                <th>Firefox</th>
                <th>Opera</th>
                <th>Microsoft Edge</th>
                <th>Safari</th>
            </tr>
            </thead>
            <tbody>
            <tr>
                <td>Linux</td>
                <td>Ubuntu 14-18</td>
                <td>78-63</td>
                <td>70-55</td>
                <td>64-49</td>
                <td>n/a</td>
                <td>n/a</td>
            </tr>
            <tr>
                <td>Windows</td>
                <td>10</td>
                <td>78-63</td>
                <td>70-55</td>
                <td>64-49</td>
                <td>18-17 Partially</td>
                <td>n/a</td>
            </tr>
            <tr>
                <td>MacOS</td>
                <td>Mojave (10.14)</td>
                <td>78-69</td>
                <td>70-62</td>
                <td>64-56</td>
                <td>n/a</td>
                <td>12</td>
            </tr>
            </tbody>
        </table>
    </div>


</div>